code/migration/src/main/scala/com/namely/chiefofstate/migration/versions/v2/V2.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 Namely Inc.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+package com.namely.chiefofstate.migration.versions.v2
+
+import akka.actor.typed.ActorSystem
+import akka.projection.slick.SlickProjection
+import akka.serialization.{Serialization, SerializationExtension}
+import com.namely.chiefofstate.migration.Version
+import org.slf4j.{Logger, LoggerFactory}
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIO
+import slick.jdbc.JdbcProfile
+
+import scala.concurrent.{Await, ExecutionContextExecutor}
+import scala.concurrent.duration.Duration
+import scala.util.{Success, Try}
+
+/**
+ * V2 migration moves the legacy journal data into the new journal
+ *
+ * @param journalJdbcConfig the journal configuration
+ * @param projectionJdbcConfig the projection configuration
+ * @param system the actor system
+ */
+case class V2(journalJdbcConfig: DatabaseConfig[JdbcProfile], projectionJdbcConfig: DatabaseConfig[JdbcProfile])(
+  implicit system: ActorSystem[_]
+) extends Version {
+  implicit val ec: ExecutionContextExecutor = system.executionContext
+  final val log: Logger = LoggerFactory.getLogger(getClass)
+
+  override def versionNumber: Int = 2
+
+  /**
+   * implement this method to upgrade the application to this version. This is
+   * run in the same db transaction that commits the version number to the
+   * database.
+   *
+   * @return a DBIO that runs this upgrade
+   */
+  override def upgrade(): DBIO[Unit] = {
+    log.info(s"finalizing ChiefOfState migration: #$versionNumber")
+    SchemasUtil.dropLegacyJournalTablesStmt.flatMap(_ => DBIO.successful {})
+  }
+
+  /**
+   * implement this method to snapshot this version (run if no prior versions found)
+   *
+   * @return a DBIO that creates the version snapshot
+   */
+  override def snapshot(): DBIO[Unit] = {
+    log.info("creating new ChiefOfState journal tables")
+    SchemasUtil.createJournalTables(journalJdbcConfig)
+
+    log.info("creating ChiefOfState read side offset tables")
+    Await.result(SlickProjection.createOffsetTableIfNotExists(projectionJdbcConfig), Duration.Inf)
```

Comment:
Instead of the infinite wait, why not convert this future into a DBIO action using [DBIO.from](https://scala-slick.org/doc/3.2.0/dbio.html#primitives)

Reply:
It is intentional because we need to make sure that the migration code run smoothly before COS start. The simplest and stupid way is to make sure all db calls execute to completion.